examples/pytorch/gcn.py

```diff
@@ -0,0 +1,214 @@
+            [
+                GCN2Conv(
+                    hidden_channels,
+                    alpha,
+                    theta,
+                    layer + 1,
+                    shared_weights,
+                    normalize=False,
+                )
+                for layer in range(num_layers)
+            ]
+        )
+
+    def forward(self, x: torch.Tensor, adj: SparseTensor):
+        """Forward pass."""
+        x = F.dropout(x, self.dropout, training=self.training)
+        x = x_0 = self.lins[0](x).relu()
+
+        for conv in self.convs:
+            x = F.dropout(x, self.dropout, training=self.training)
+            x = conv(x, x_0, adj)
+            x = x.relu()
+
+        x = F.dropout(x, self.dropout, training=self.training)
+        x = self.lins[1](x)
+
+        return x.log_softmax(dim=-1)
+
+
+def _train_func(config: dict):
+    query_param = GCNQueryParameter(
+        neighbor_edge_types=np.array(
+            [0], np.int32
+        ),  # edges to nodes in test datasets usually have type 1, but we can use 0 for Cora.
+        **config["graph_query"],
+    )
+    model = GCN(
+        feature_dim=query_param.feature_dim,
+        num_classes=query_param.num_classes,
+        hidden_channels=64,
+        num_layers=64,
+        alpha=0.1,
+        theta=0.5,
+        shared_weights=True,
+        dropout=0.6,
+    ).to(config["device"])
+
+    optimizer = torch.optim.Adam(
+        [
+            dict(params=model.convs.parameters(), weight_decay=0.01),
+            dict(params=model.lins.parameters(), weight_decay=5e-4),
+        ],
+        lr=0.01,
+    )
+
+    g = config["graph"]
+    masks = {}
+    for mask in ["train", "test"]:
+        full_set = torch.zeros(g.NUM_NODES, dtype=torch.bool)
+        nodes = torch.from_numpy(
+            np.loadtxt(f"{config['data_dir']}/{mask}.nodes", dtype=np.int64)
+        )
+        full_set[nodes] = True
+        masks[mask] = full_set
+
+    # GCN doesn't need any sampling, so we can just query the full graph.
+    feat, adj, labels = query(np.arange(g.NUM_NODES, dtype=np.int64), g, query_param)
+    accuracy: float = 0
+
+    for epoch in range(config["num_epochs"]):
+        total_loss = 0
+        num_examples = 0
+        model.train()
+        optimizer.zero_grad()
+        out = model(feat, adj)
+        loss = F.nll_loss(out[masks["train"]], labels[masks["train"]])
+        loss.backward()
+        optimizer.step()
+        num_examples += len(labels[masks["train"]])
+        total_loss += loss.item() * len(labels[masks["train"]])
+        if epoch % 10 == 0:
+            model.eval()
+            ys = labels[masks["test"]]
+            preds = out[masks["test"]].argmax(dim=-1).cpu()
+            accuracy = int((ys == preds).sum()) / int(len(ys))
+            get_logger().info(
+                f"Epoch: {epoch:>2d}; Loss: {(total_loss/num_examples):>4.2f}; Accuracy: {accuracy:.4f};"
+            )
+
+    assert accuracy > 0.45
```

Comment:
Is 45% accuracy the best we can do?

Reply:
No, this is a good enough number for training for 100 epochs under 30 seconds.